### PR TITLE
Migrate deployment controller log messages to structured logging

### PR DESCRIPTION
**Pull request by aldudko**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:
Ref:
* [keps/sig-instrumentation/1602-structured-logging](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging)
* [Structured Logging migration instructions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md)

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Verify log output**:

```
# before

deployment_controller.go:153] Starting deployment controller
deployment_controller.go:169] Adding deployment coredns
deployment_controller.go:176] Updating deployment foo
deployment_controller.go:194] Deleting deployment nginx-deployment
deployment_controller.go:215] ReplicaSet rs1 added.
deployment_controller.go:226] Orphan ReplicaSet rs1 added.
deployment_controller.go:246] user error! more than one deployment is selecting replica set default/rs1 with labels: map[string]string{"foo":"bar"}, returning default/d1
deployment_controller.go:281] ReplicaSet rs1 updated.
deployment_controller.go:294] Orphan ReplicaSet rs1 updated.
deployment_controller.go:333] ReplicaSet rs2 deleted.
deployment_controller.go:357] Pod nginx-deployment-66b6c48dd5-whsv4 deleted.
deployment_controller.go:425] Cannot get replicaset "nginx-deployment-66b6c48dd5" for pod "nginx-deployment-66b6c48dd5-cn525": replicaset.apps "nginx-deployment-66b6c48dd5" not found
deployment_controller.go:490] "Error syncing deployment" deployment="kube-system/coredns" err="Operation cannot be fulfilled on deployments.apps \"coredns\": the object has been modified; please apply your changes to the latest version and try again"
deployment_controller.go:570] Started syncing deployment "kube-system/coredns" (2020-12-24 06:54:00.960340479 +0000 UTC m=+36.137736457)
deployment_controller.go:572] Finished syncing deployment "kube-system/coredns" (15.288708ms)
deployment_controller.go:581] Deployment kube-system/nginx-deployment has been deleted
```

```
# after

deployment_controller.go:153] "Starting controller" controller="deployment"
deployment_controller.go:175] "Adding deployment" deployment="kube-system/coredns"
deployment_controller.go:176] "Updating deployment" deployment="default/nginx-deployment"
deployment_controller.go:200] "Deleting deployment" deployment="kube-system/nginx-deployment"
deployment_controller.go:221] "ReplicaSet added" replicaSet="default/rs1"
deployment_controller.go:232] "Orphan ReplicaSet added" replicaSet="default/rs1"
deployment_controller.go:252] "user error! more than one deployment is selecting replica set" replicaSet="default/rs1" labels=map[foo:bar] deployment="default/d2"
deployment_controller.go:287] "ReplicaSet updated" replicaSet="default/rs1"
deployment_controller.go:300] "Orphan ReplicaSet updated" replicaSet="default/rs1"
deployment_controller.go:339] "ReplicaSet deleted" replicaSet="kube-system/nginx-deployment-66b6c48dd5"
deployment_controller.go:357] "Pod deleted" pod="default/nginx-deployment-66b6c48dd5-l97lb"
deployment_controller.go:425] "Cannot get replicaset for pod" ownerReference="nginx-deployment-66b6c48dd5" pod="default/nginx-deployment-66b6c48dd5-d5xgj" err=<nil>
deployment_controller.go:490] "Error syncing deployment" deployment="kube-system/coredns" err="Operation cannot be fulfilled on deployments.apps \"coredns\": the object has been modified; please apply your changes to the latest version and try again"
deployment_controller.go:576] "Started syncing deployment" deployment="default/foo" startTime="2020-12-24 10:38:37.774226481 -0800 PST m=+0.028639963"
deployment_controller.go:578] "Finished syncing deployment" deployment="default/foo" duration="891.282µs"
deployment_controller.go:587] "Deployment has been deleted" deployment="kube-system/nginx-deployment"
```

### CODE REVIEW

File: pkg/controller/deployment/deployment_controller.go [hunk 1]
```diff
@@ -166,14 +172,14 @@ func (dc *DeploymentController) Run(workers int, stopCh <-chan struct{}) {
 
 func (dc *DeploymentController) addDeployment(obj interface{}) {
 	d := obj.(*apps.Deployment)
-	klog.V(4).Infof("Adding deployment %s", d.Name)
+	klog.V(4).InfoS("Adding deployment", klogKeyDeployment, klog.KObj(d))
```
Comment (krmayankk) [inline pkg/controller/deployment/deployment_controller.go:175]: i see repetition for deployment in `"Adding deployment" deployment="kube-system/coredns"`

This can simply be `"Adding" deployment="kube-system/coredns"`

File: pkg/controller/deployment/deployment_controller.go [hunk 2]
```diff
@@ -191,7 +197,7 @@ func (dc *DeploymentController) deleteDeployment(obj interface{}) {
 			return
 		}
 	}
-	klog.V(4).Infof("Deleting deployment %s", d.Name)
+	klog.V(4).InfoS("Deleting deployment", klogKeyDeployment, klog.KObj(d))
```
Comment (krmayankk) [inline pkg/controller/deployment/deployment_controller.go:200]: similar, we can remove the duplication of `deployment`

File: pkg/controller/deployment/deployment_controller.go [hunk 3]
```diff
@@ -223,7 +229,7 @@ func (dc *DeploymentController) addReplicaSet(obj interface{}) {
 	if len(ds) == 0 {
 		return
 	}
-	klog.V(4).Infof("Orphan ReplicaSet %s added.", rs.Name)
+	klog.V(4).InfoS("Orphan ReplicaSet added", klogKeyReplicaSet, klog.KObj(rs))
```
Comment (krmayankk) [inline pkg/controller/deployment/deployment_controller.go:232]: `Orphan added"

File: pkg/controller/deployment/deployment_controller.go [hunk 4]
```diff
@@ -278,7 +284,7 @@ func (dc *DeploymentController) updateReplicaSet(old, cur interface{}) {
 		if d == nil {
 			return
 		}
-		klog.V(4).Infof("ReplicaSet %s updated.", curRS.Name)
+		klog.V(4).InfoS("ReplicaSet updated", klogKeyReplicaSet, klog.KObj(curRS))
```
Comment (krmayankk) [inline pkg/controller/deployment/deployment_controller.go:287]: `updated`

File: pkg/controller/deployment/deployment_controller.go [hunk 5]
```diff
@@ -567,9 +573,9 @@ func (dc *DeploymentController) getPodMapForDeployment(d *apps.Deployment, rsLis
 // This function is not meant to be invoked concurrently with the same key.
 func (dc *DeploymentController) syncDeployment(key string) error {
 	startTime := time.Now()
-	klog.V(4).Infof("Started syncing deployment %q (%v)", key, startTime)
+	klog.V(4).InfoS("Started syncing deployment", klogKeyDeployment, key, "startTime", startTime)
 	defer func() {
-		klog.V(4).Infof("Finished syncing deployment %q (%v)", key, time.Since(startTime))
+		klog.V(4).InfoS("Finished syncing deployment", klogKeyDeployment, key, "duration", time.Since(startTime))
```
Comment (krmayankk) [inline pkg/controller/deployment/deployment_controller.go:578]: "Finshed syncing"

**Review by krmayankk — COMMENTED** [review]
added some suggestions, ptal

File: pkg/controller/deployment/deployment_controller.go (hunk 1 above)
Comment (aldudko) [reply pkg/controller/deployment/deployment_controller.go:175]: Using complete sentences in log message is in line with [Structured Logging migration instructions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md) and [basic examples in the klog package](https://github.com/kubernetes/klog/blob/master/klog.go#L1409).
```
// >> klog.InfoS("Pod status updated", "pod", "kubedns", "status", "ready")
// output:
// >> I1025 00:15:15.525108       1 controller_utils.go:116] "Pod status updated" pod="kubedns" status="ready"
```

That repetition is accidental. It occurs because I use `deployment` as a keyword.
If I used something else instead of `deployment` we would not see any repetition here. For example:
```
klog.V(4).InfoS("Adding deployment", "id", klog.KObj(d))
// "Adding deployment" id="kube-system/coredns"
```

On the opposite side reducing the log message to `Adding` would lead to an incomplete message which wouldn't explain what exactly we are adding. For example:
```
klog.V(4).InfoS("Adding", "id", klog.KObj(d))
// "Adding" id="kube-system/coredns"
```

Another disadvantage of using standalone `Adding` is that when there are more than 1 key/value pair provided, it creates ambiguity because it is not clear which of the keywords `Adding` is referring to. It is not guaranteed that downstream libraries preserve keyword ordering. 

```
klog.V(4).InfoS("Adding", "pod", klog.KObj(pod), "deployment", klog.KObj(deployment), "replicaSet", klog.KObj(replicaSet))
// "Adding" pod="kube-system/coredns-xxxx" deployment="kube-system/coredns-xxxx" replicaSet="kube-system/coredns-xxxx"
// or
// "Adding" deployment="kube-system/coredns-xxxx" replicaSet="kube-system/coredns-xxxx" pod="kube-system/coredns-xxxx"
// or
// "Adding" replicaSet="kube-system/coredns-xxxx" deployment="kube-system/coredns-xxxx" pod="kube-system/coredns-xxxx"
```

That's why I believe keeping the subject in the message (i.e. `Adding deployment`) is required.

Let me know your thoughts please.

File: pkg/controller/deployment/deployment_controller.go (hunk 2 above)
Comment (aldudko) [reply pkg/controller/deployment/deployment_controller.go:200]: same as above (line 175)

File: pkg/controller/deployment/deployment_controller.go (hunk 3 above)
Comment (aldudko) [reply pkg/controller/deployment/deployment_controller.go:232]: same as above (line 175)

File: pkg/controller/deployment/deployment_controller.go (hunk 4 above)
Comment (aldudko) [reply pkg/controller/deployment/deployment_controller.go:287]: same as above (line 175)

File: pkg/controller/deployment/deployment_controller.go (hunk 5 above)
Comment (aldudko) [reply pkg/controller/deployment/deployment_controller.go:578]: same as above (line 175)

File: pkg/controller/deployment/deployment_controller.go (hunk 1 above)
Comment (krmayankk) [reply pkg/controller/deployment/deployment_controller.go:175]: how about we change the key to be simply `name` ? then we will have `"Adding deployment" name="kube-system/coredns"` wdyt ?

Comment (aldudko) [reply pkg/controller/deployment/deployment_controller.go:175]: I like `name`.
However, I believe we should be consistent across different components.
If we use `name` here, we should be using `name` for pod names, statefulSet names, etc.
Otherwise, log readers will be having a hard time trying to figure out which keyword they need to use to get the logs they are interested in.

Let me ping someone from #sig-instrumentation who drives this migration so that they share their vision.
Personally, I don't have a strong opinion here.

Comment (serathius) [reply pkg/controller/deployment/deployment_controller.go:175]: Hey, author of migration instructions here. Based on current guidelines most correct version would be `"Adding deployment" deployment="kube-system/coredns"`. Some arguments:
* This log will include repetition, but this repetition is required to make meaning of both message and it's metadata independent. Message alone should be able to provide good enough description of event that is happening. 
* We should try to use consistent key names that are depended on context of any specific log. This is with goal of being able to answer queries like "Give me all logs relating to this specific deployment". Meaning of key `name` proposed above is based on context of provided by message which is not structured, so in this log is stores deployment reference, but in next one it could store pod name, making it not very useful for correlating log lines.

File: pkg/controller/deployment/deployment_controller.go [hunk 6]
```diff
@@ -487,13 +493,13 @@ func (dc *DeploymentController) handleErr(err error, key interface{}) {
 	}
 
 	if dc.queue.NumRequeues(key) < maxRetries {
-		klog.V(2).InfoS("Error syncing deployment", "deployment", klog.KRef(ns, name), "err", err)
+		klog.V(2).InfoS("Error syncing deployment", klogKeyDeployment, klog.KRef(ns, name), "key", key, "err", err)
```
Comment (serathius) [inline pkg/controller/deployment/deployment_controller.go:496]: why introduce "key" key here? Based on name and log message it's impossible to interpret what value is stored under it without looking at the source code. I would omit  this key as this log message is not related to cache key at all.

File: pkg/controller/deployment/deployment_controller.go [hunk 7]
```diff
@@ -487,13 +493,13 @@ func (dc *DeploymentController) handleErr(err error, key interface{}) {
...
 		dc.queue.AddRateLimited(key)
 		return
 	}
 
 	utilruntime.HandleError(err)
-	klog.V(2).Infof("Dropping deployment %q out of the queue: %v", key, err)
+	klog.V(2).InfoS("Dropping deployment out of the queue", klogKeyDeployment, klog.KRef(ns, name), "key", key, "err", err)
```
Comment (serathius) [inline pkg/controller/deployment/deployment_controller.go:502]: ditto

File: pkg/controller/deployment/deployment_controller.go [hunk 8]
```diff
@@ -60,6 +60,12 @@ const (
 	maxRetries = 15
 )
 
+const (
+	// Structured Logging keys
+	klogKeyDeployment = "deployment"
```
Comment (serathius) [inline pkg/controller/deployment/deployment_controller.go:65]: I would say that using raw key names instead of extracting const for logging is one example of justified breaking of "do not repeat yourself" rule. 

For this PR we can leave it as it is, but I would prefer avoiding creating pattern. Will try to add a section in migration guide about this.

File: pkg/controller/deployment/deployment_controller.go [hunk 9]
```diff
@@ -150,8 +156,8 @@ func (dc *DeploymentController) Run(workers int, stopCh <-chan struct{}) {
 	defer utilruntime.HandleCrash()
 	defer dc.queue.ShutDown()
 
-	klog.Infof("Starting deployment controller")
-	defer klog.Infof("Shutting down deployment controller")
+	klog.InfoS("Starting deployment controller")
+	defer klog.InfoS("Shutting down deployment controller")
```
[the remark this answers is not in the document]
Comment (serathius) [reply pkg/controller/deployment/deployment_controller.go:160]: I think this is first controller migrated,

File: pkg/controller/deployment/deployment_controller.go (hunk 6 above)
Comment (aldudko) [reply pkg/controller/deployment/deployment_controller.go:496]: when `cache.SplitMetaNamespaceKey` in [line 484](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/deployment_controller.go#L484)  returns an error, both `ns` and `name` are empty strings.
The message we get here in such situation will looks like that: Error syncing deployment deployment=""
`key` was already introduced in line 486, and I reused it to provide more information when such problem occurs.

Another options I can think of is to use `key.(string)` as deployment name instead of klog.KRef(ns, name) if `keyErr` != nil.
Let me know please if you prefer to go this direction.

File: pkg/controller/deployment/deployment_controller.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (aldudko) [reply pkg/controller/deployment/deployment_controller.go:160]: replaced with
```
       klog.InfoS("Starting controller", "controller", "deployment")
       defer klog.InfoS("Shutting down controller", "controller", "deployment")
```

File: pkg/controller/deployment/deployment_controller.go (hunk 8 above)
Comment (aldudko) [reply pkg/controller/deployment/deployment_controller.go]: "do not repeat yourself" was not the reason I introduced the constants.

There is no way to validate that everyone uses the same keys for the same objects.
Someone can write "replicaSet", someone else can write ""replicaset" or make a typo.
The constants could be a way of ensuring the key consistency.

Anyway, I wasn't sure about that in the first place and wanted to get feedback on that.
Removed the constants for now.

File: pkg/controller/deployment/deployment_controller.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (serathius) [reply pkg/controller/deployment/deployment_controller.go]: LGTM

File: pkg/controller/deployment/deployment_controller.go (hunk 8 above)
Comment (serathius) [reply pkg/controller/deployment/deployment_controller.go]: You mentioned important problem: typos in key names.

I don't think we should solve it via consts, but we should look into static analysis. 
Main problem with consts is that they don't scale. To standardize key naming we would need to have one global repository of consts for whole k/k repo. This is not a dependency that we want to maintain, also it makes it even harder for projects outside k/k.

For now we don't want to enforce closed set of keys to use, but we should prevent typos like you mentioned. Those simple cases should be easy to detect without introducing to much restrictions on key names. We could add simple `hack/verify` script that prevents merging code with typos.

After one recent review where most issues where 3-4 different ways to write `podSandboxID`. I think we should add it to Beta scope. Possibly we can extend static analysis piece we already planned.
/cc @ehashman

File: pkg/controller/deployment/deployment_controller.go [hunk 10]
```diff
@@ -487,13 +487,13 @@ func (dc *DeploymentController) handleErr(err error, key interface{}) {
 	}
 
 	if dc.queue.NumRequeues(key) < maxRetries {
-		klog.V(2).InfoS("Error syncing deployment", "deployment", klog.KRef(ns, name), "err", err)
+		klog.V(2).InfoS("Error syncing deployment", "deployment", klog.KRef(ns, name), "key", key, "err", err)
```
Comment (serathius) [inline pkg/controller/deployment/deployment_controller.go:490]: `key` can be inferred from `deployment` making it redundant. 
```suggestion
		klog.V(2).InfoS("Error syncing deployment", "deployment", klog.KRef(ns, name), "err", err)
```

File: pkg/controller/deployment/deployment_controller.go [hunk 11]
```diff
@@ -567,9 +567,9 @@ func (dc *DeploymentController) getPodMapForDeployment(d *apps.Deployment, rsLis
 // This function is not meant to be invoked concurrently with the same key.
 func (dc *DeploymentController) syncDeployment(key string) error {
 	startTime := time.Now()
-	klog.V(4).Infof("Started syncing deployment %q (%v)", key, startTime)
+	klog.V(4).InfoS("Started syncing deployment", "deployment", key, "startTime", startTime)
```
Comment (serathius) [inline pkg/controller/deployment/deployment_controller.go:570]: We should avoid passing different values into same logging key, `deployment` is reserved for using `klog.KObj` or `klog.KRef`. I would propose to either

Parse key and use KRef

```suggestion
    ns, name, _ := cache.SplitMetaNamespaceKey(key.(string))
	klog.V(4).InfoS("Started syncing deployment", "deployment", klog.KRef(ns, name), "startTime", startTime)
```

or use separate key
```suggestion
	klog.V(4).InfoS("Started syncing deployment matching cache key", "cacheKey", key, "startTime", startTime)
```

First method is more consistent with other logs, but results in more complicated code.
Second option is easier to implement, but less useful for correlating logs.

Comment (serathius) [reply pkg/controller/deployment/deployment_controller.go:570]: Noticed code below, looks like key is parsed couple of lines below. We could change order of lines  too.

File: pkg/controller/deployment/deployment_controller.go [hunk 12]
```diff
@@ -567,9 +567,9 @@ func (dc *DeploymentController) getPodMapForDeployment(d *apps.Deployment, rsLis
...
 	defer func() {
-		klog.V(4).Infof("Finished syncing deployment %q (%v)", key, time.Since(startTime))
+		klog.V(4).InfoS("Finished syncing deployment", "deployment", key, "duration", time.Since(startTime))
```
Comment (serathius) [inline pkg/controller/deployment/deployment_controller.go:572]: ditto with `key`

File: pkg/controller/deployment/deployment_controller.go (hunk 11 above)
Comment (aldudko) [reply pkg/controller/deployment/deployment_controller.go]: changed the order of lines in favor of the first (more consistent method) and replaced key with `klog.KRef(ns, name)`.
Thanks!

File: pkg/controller/deployment/deployment_controller.go (hunk 12 above)
Comment (aldudko) [reply pkg/controller/deployment/deployment_controller.go]: replaced `key` by `klog.KRef(namespace, name)`.

File: pkg/controller/deployment/deployment_controller.go (hunk 10 above)
Comment (aldudko) [reply pkg/controller/deployment/deployment_controller.go]: removed `key`.

File: pkg/controller/deployment/deployment_controller.go (hunk 7 above)
Comment (aldudko) [reply pkg/controller/deployment/deployment_controller.go]: removed `key`.

File: pkg/controller/deployment/deployment_controller.go (hunk 6 above)
Comment (aldudko) [reply pkg/controller/deployment/deployment_controller.go]: removed `key`.

File: pkg/controller/deployment/deployment_controller.go [hunk 13]
```diff
@@ -566,19 +566,21 @@ func (dc *DeploymentController) getPodMapForDeployment(d *apps.Deployment, rsLis
...
 func (dc *DeploymentController) syncDeployment(key string) error {
-	startTime := time.Now()
-	klog.V(4).Infof("Started syncing deployment %q (%v)", key, startTime)
-	defer func() {
-		klog.V(4).Infof("Finished syncing deployment %q (%v)", key, time.Since(startTime))
-	}()
-
 	namespace, name, err := cache.SplitMetaNamespaceKey(key)
 	if err != nil {
+		klog.ErrorS(err, "Failed to split meta namespace cache key", "key", key)
```
Comment (serathius) [inline pkg/controller/deployment/deployment_controller.go:571]:
```suggestion
		klog.ErrorS(err, "Failed to split meta namespace cache key", "cacheKey", key)
```

Comment (aldudko) [reply pkg/controller/deployment/deployment_controller.go]: oh, right.
Updated here and above where I copy-pasted that from.

File: pkg/controller/deployment/deployment_controller.go [hunk 14]
```diff
@@ -150,8 +150,8 @@ func (dc *DeploymentController) Run(workers int, stopCh <-chan struct{}) {
 	defer utilruntime.HandleCrash()
 	defer dc.queue.ShutDown()
 
-	klog.Infof("Starting deployment controller")
-	defer klog.Infof("Shutting down deployment controller")
+	klog.InfoS("Starting controller", "controller", "deployment")
+	defer klog.InfoS("Shutting down controller", "controller", "deployment")
```
Comment (soltysh) [inline pkg/controller/deployment/deployment_controller.go:154]: I don't understand the above changes, there are no variables here so I'd leave it as is.

File: pkg/controller/deployment/deployment_controller.go [hunk 15]
```diff
@@ -566,19 +566,21 @@ func (dc *DeploymentController) getPodMapForDeployment(d *apps.Deployment, rsLis
...
 	if err != nil {
+		klog.ErrorS(err, "Failed to split meta namespace cache key", "cacheKey", key)
 		return err
 	}
+
+	startTime := time.Now()
+	klog.V(4).InfoS("Started syncing deployment", "deployment", klog.KRef(namespace, name), "startTime", startTime)
+	defer func() {
+		klog.V(4).InfoS("Finished syncing deployment", "deployment", klog.KRef(namespace, name), "duration", time.Since(startTime))
+	}()
```
Comment (soltysh) [inline pkg/controller/deployment/deployment_controller.go:579]: Why moving this?

**Review by soltysh — COMMENTED** [review]
Few nits

File: pkg/controller/deployment/deployment_controller.go (hunk 15 above)
Comment (aldudko) [reply pkg/controller/deployment/deployment_controller.go:579]: per https://github.com/kubernetes/kubernetes/pull/97507#discussion_r575173518 `We should avoid passing different values into same logging key, deployment is reserved for using klog.KObj or klog.KRef`.

I moved those line to be able to use `klog.KRef(namespace, name)` to write deployment name instead of cacheKey.

File: pkg/controller/deployment/deployment_controller.go (hunk 14 above)
Comment (aldudko) [reply pkg/controller/deployment/deployment_controller.go:154]: We don't keep it in a variable, but "deployment" is a controller name, isn't it?

If it was a pod, we would convert something like `"Starting pod ABC"` to `"Starting pod", pod="ABC"`.
I apply the same logic to the controller name. 

Also, If we don't have any keys, the only way to find when did the deployment controller start is to use full text search and look for "Starting deployment  controller". While specifying `controller=deployment` in a log viewer gives us a quick way to get this information.

Comment (soltysh) [reply pkg/controller/deployment/deployment_controller.go:154]: That's fair.

File: pkg/controller/deployment/deployment_controller.go (hunk 15 above)
Comment (soltysh) [reply pkg/controller/deployment/deployment_controller.go:579]: kk

**Review by soltysh — APPROVED** [review]
/lgtm
/approve
/ok-to-test
/triage accepted
/priority backlog
